quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs

```diff
@@ -386,18 +458,28 @@ runActions (Actions_ rejected (Smart _ actions)) = loop initialAnnotatedState []
     loop :: Annotated state -> Env m -> [Step state] -> PropertyM m (Annotated state, Env m)
     loop _s env [] = do
       unless (null rejected) $
-        monitor (tabulate "Actions rejected by precondition" rejected)
+        monitor $
+          tabulate "Actions rejected by precondition" rejected
       return (_s, reverse env)
     loop s env ((v := act) : as) = do
       pre $ computePrecondition s act
-      ret <- run (perform (underlyingState s) act (lookUpVar env))
-      let name = actionName act
-      monitor (tabulate "Actions" [name])
+      ret <- run $ perform (underlyingState s) (polarAction act) (lookUpVar env)
+      let name = show (polarity act) ++ actionName (polarAction act)
+      monitor $ tabulate "Actions" [name]
       let var = unsafeCoerceVar v
           s' = computeNextState s act var
           env' = (var :== ret) : env
-      monitor (monitoring @state @m (underlyingState s, underlyingState s') act (lookUpVar env') ret)
-      (b, (Endo mon, Endo onFail)) <- run . runWriterT . runPost $ postcondition @state @m (underlyingState s, underlyingState s') act (lookUpVar env) ret
+      monitor $ tabulate "Action polarity" [show $ polarity act]
+      monitor $ monitoring @state @m (underlyingState s, underlyingState s') (polarAction act) (lookUpVar env') ret
+      (b, (Endo mon, Endo onFail)) <-
```

Comment:
😓

Reply:
That's fourmolol I'm afraid 😅

Reply:
Ah yes, and the expression itself is somewhat involved.